### PR TITLE
refactor train_spec.loss_fn to build_loss_fn & chunked loss

### PR DESCRIPTION
**Pull request by lkhphuc**
To enable chunked loss, I refactor the `train_spec.loss_fn` to `train_spec.build_loss_fn` as suggested in #996

 Also:
- Unified chunked loss implementation with the `cross_entropy_loss` name, since default is a special case with chunk=1.
Identical loss and memory:

Screenshot 2025-03-30 at 13 03 44

- Enable compiling the loss function as it's working fine for me, and result in reduction of memory and slightly better througput.

Screenshot 2025-03-30 at 16 47 00

While we only have one type of loss, I create a field for `loss.name` with default to 'cross_entropy' to be fork-friendly anyway.

### CODE REVIEW

File: torchtitan/components/loss.py [hunk 1]
```diff
@@ -4,32 +4,45 @@
...
-
-    # compute one chunk at a time
     total_loss = 0.0
-    for logits_chunk, labels_chunk in zip(logits, labels):
-        total_loss += cross_entropy_loss(logits_chunk, labels_chunk)
+    for logits_chunk, labels_chunk in zip(
+        logits.chunk(num_output_chunks, dim=1),
+        labels.chunk(num_output_chunks, dim=1),
+        strict=True,
+    ):
```
Comment (tianyu-l) [inline torchtitan/components/loss.py]: There might be overhead associated with `chunk` and `zip`. Can we early return non-chunk version if `num_output_chunks==1`?

File: torchtitan/config_manager.py [hunk 2]
```diff
@@ -226,6 +226,27 @@ def __init__(self):
...
+            "--loss.name",
+            type=str,
+            default="cross_entropy",
+            choices=["cross_entropy"],
+            help=""" The name of the loss function to use.
+            Currently only support (chunked) cross entropy loss for LLM training.
+            """,
+        )
+        self.parser.add_argument(
+            "--loss.num_chunks",
```
Comment (tianyu-l) [inline torchtitan/config_manager.py]: I'd imagine not all loss functions would support `num_chunks` sequential computation, but it's OK to keep this field until change is needed.

File: torchtitan/config_manager.py [hunk 3]
```diff
@@ -226,6 +226,27 @@ def __init__(self):
             """,
         )
 
+        # loss configs
+        self.parser.add_argument(
+            "--loss.name",
+            type=str,
+            default="cross_entropy",
+            choices=["cross_entropy"],
+            help=""" The name of the loss function to use.
```
Comment (tianyu-l) [inline torchtitan/config_manager.py]:
```suggestion
            help="""The name of the loss function to use.
```
not sure how space works, maybe nit

File: torchtitan/config_manager.py [hunk 4]
```diff
@@ -226,6 +226,27 @@ def __init__(self):
...
+            Currently only support (chunked) cross entropy loss for LLM training.
+            """,
+        )
+        self.parser.add_argument(
+            "--loss.num_chunks",
+            type=int,
+            default=1,
+            help=""" Number of chunks to split the logits into and sequentially
```
Comment (tianyu-l) [inline torchtitan/config_manager.py]:
```suggestion
            help="""Number of chunks to split the logits into and sequentially
```

File: torchtitan/models/llama/train_configs/debug_model.toml [hunk 5]
```diff
@@ -28,6 +28,9 @@ norm_type = "rmsnorm"  # layernorm / np_layernorm / rmsnorm
 tokenizer_path = "./tests/assets/test_tiktoken.model"
 # converters = "float8"
 
+[loss]
+# num_chunks = 8  # Reduce peak memory usage
```
Comment (tianyu-l) [inline torchtitan/models/llama/train_configs/debug_model.toml]: I think keeping this commented out in `debug_model.toml` is good enough. Maybe let's remove the field in other toml.

File: torchtitan/protocols/train_spec.py [hunk 6]
```diff
@@ -60,6 +60,7 @@ def from_model_args(cls, args: BaseModelArgs) -> nn.Module:
 ]
 LRSchedulersBuilder: TypeAlias = Callable[[OptimizersContainer], LRSchedulersContainer]
 LossFunction: TypeAlias = Callable[[torch.Tensor, torch.Tensor], torch.Tensor]
```
Comment (tianyu-l) [inline torchtitan/protocols/train_spec.py]: Do we ever use this directly other than the line below?
I think we should either move the `LossFunction` to `loss.py` and import here (like `OptimizerContainer`), or remove it entirely (and use `Callable[...` as type directly).

File: torchtitan/train.py [hunk 7]
```diff
@@ -154,6 +154,8 @@ def __init__(self, job_config: JobConfig):
         model_converters = build_model_converters(job_config, parallel_dims)
         model_converters.convert(model)
 
+        self.loss_fn = self.train_spec.build_loss_fn(job_config)
```
Comment (tianyu-l) [inline torchtitan/train.py]: 1. please add an explicit class variable `loss_fn` before `Trainer.__init__`
2. please move this assignment to where it is going to be used for the first time, i.e. before PP uses it.

**Review by tianyu-l — COMMENTED** [review]
This makes sense to me in general. Please address comments. Thanks!

**Comment by tianyu-l** [comment]
Hmm two more questions:
1. From the picture, why chunked loss and original loss are not identical? Did you fix seed and determinism and compare?
2. Good to know that compiled loss works again. I saw it has lowest memory, even compared with chunked loss and compiled chunked loss. Is there motivation to have chunked loss anyway?

**Comment by lkhphuc** [comment]
1. I set determinisic and seed and get identical loss for chunk=1 and original loss.

Screenshot 2025-03-31 at 13 54 51

2. I tested more thoroughly on longer sequence (16k) and bigger batch size. It's true that compiled saved more memory than chunked loss ever could (it plateaued as 16 chunks).
If chunked >1 with compile, it does take slightly more memory than compile with 1 chunk. 
I don't see a reason to use chunked loss with compile either.
Screenshot 2025-03-31 at 14 33 19

Wdyt? I could remove the chunk loss altogether or just leave it there with default =1 and a note/warning about memory usage together with compile.

**Review by tianyu-l — COMMENTED** [review]
At this point I don't see great value from chunked cross entropy loss, compared with compiled loss. So I'm leaning towards removing it all.

I think we can keep the `build_loss_fn` abstraction, and deciding if we use compile from the `job_config` part. WDYT?

**Comment by lkhphuc** [comment]
Ok I removed the chunked loss. Now the loss fn builder has only 1 responsibility of compiling the loss fn if `job_config.training.compile` is set.

File: torchtitan/components/loss.py [hunk 8]
```diff
@@ -14,22 +20,9 @@ def cross_entropy_loss(pred: torch.Tensor, labels: torch.Tensor) -> torch.Tensor
...
-    labels = [target_chunk for target_chunk in labels.chunk(num_output_chunks, dim=1)]
-    logits = [logit_chunk for logit_chunk in logits.chunk(num_output_chunks, dim=1)]
-
-    # compute one chunk at a time
-    total_loss = 0.0
-    for logits_chunk, labels_chunk in zip(logits, labels):
-        total_loss += cross_entropy_loss(logits_chunk, labels_chunk)
-
-    return total_loss / num_output_chunks
+def build_cross_entropy_loss_fn(job_config: JobConfig):
```
Comment (tianyu-l) [inline torchtitan/components/loss.py]: nit
```suggestion
def build_cross_entropy_loss(job_config: JobConfig):
```

**Review by tianyu-l — APPROVED** [review]
LGTM, thank you!
Had one nit comment on naming

**Review by tianyu-l — CHANGES_REQUESTED** [review]
would you please fix all the CI failures? thanks!
https://github.com/pytorch/torchtitan/actions/runs/14196589625
https://github.com/pytorch/torchtitan/actions/runs/14196589631
https://github.com/pytorch/torchtitan/actions/runs/14196589660

**Comment by lkhphuc** [comment]
Oops forgot to change the names at other places. Should all pass now.

**Comment by casper-hansen** [comment]
> At this point I don't see great value from chunked cross entropy loss, compared with compiled loss. So I'm leaning towards removing it all.
> 
> I think we can keep the `build_loss_fn` abstraction, and deciding if we use compile from the `job_config` part. WDYT?

I think the chunked cross entropy loss was removed by mistake here. The original documentation says the following, but you were compiling the entire function. There are possibly more memory savings to be had with compile.

*When using this class, we recommend using :func:`torch.compile` only on the method ``compute_cross_entropy``.
    The gains from chunking won't be realized if you compile the entire class.*

https://github.com/pytorch/torchtune/blob/main/torchtune/modules/loss/ce_chunked_output_loss.py#L24-L25

**Comment by lkhphuc** [comment]
Wasn't discussed here, but I did tried compiling the inner, per-chunked, loss fn but the memory usage is still higher than compile with single chunk. 

Maybe I did it wrong and somebody can verify?
